README.md
# RecordGalleryCard

[![npm package][npm-badge]][npm]

Used for displaying a record as a gallery card.	

![RecordGalleryCard](https://user-images.githubusercontent.com/44947294/48668890-51e64780-eb2a-11e8-8ee8-152d780be21e.png)

## Getting started

````
npm install @cmds/record-gallery-card --save
````

### Prop Types

| Property | Type | Required? | Description |
|:---|:---|:---:|:---|
| id | String | ✓ | Unique id for the instance of this record |
| name | String | ✓ | Name for the record |
| coverAttachments | Array | | Each attachment be displayed in a the card's cover as a carousel |
| coverFitTypeId | CoverFitType | | Either `crop` or `fit`. The images inside the cover will be either covered or contained, respectively |
| coverEnabled | Boolean | ✓ | Whether the cover should be displayed or not |
| fields | Array | ✓ | List of field objects containing the name and more information for each field. |
| visibleFieldOrder | Array | ✓ | A list of ids for the fields that need to be displayed and in which order |
| fieldHeightGetter | Function | ✓ | Responsible for returning the height for the field: `({field: object}): number` |
| fieldRenderer | Function | ✓ | Responsible for rendering a field given it's configuration: `({recordId: string, index: number, field: object}): jsx` [Learn more](#fieldRenderer) |

#### fieldRenderer

Responsible for rendering a field given it's configuration.

```jsx harmony
import SingleLineTextField from '@cmds/single-line-text-field'
// import all other fields that need to be supported...

const renderers = {
    singleLineText: ({props}) => (
        <SingleLineTextField
            {...props}
            text={'Luke Skywalker'}
            onChange={({id, text}) => {
                
                // store new value
            }}
        />
    ),
    // and all other fields that need to be supported
}

function fieldRenderer({id, index, field, props}) {

    const renderer = renderers[field.typeId]
    
    if (!renderer) {
        throw new Error(`Renderer for typeId '${field.typeId}' not found`)
    }
    
    /**
     * Note — props already contains properties
     * related to the context in which the field
     * gets rendered.
     * 
     * {
     *      id: 'fld1', // the field's id
     *      contextId: 'recordGalleryCard',
     *      roleId: 'readOnly'
     * }
     */
    
    return renderer({ 
        id, 
        field,
        props
    })
}
```

### More information

This component is designed and developed as part of [Cosmos Design System][cmds]. 

[cmds]: https://github.com/entercosmos/cosmos
[npm-badge]: https://img.shields.io/npm/v/@cmds/record-gallery-card.svg
[npm]: https://www.npmjs.org/package/@cmds/record-gallery-card


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.8.1] - 2018-11-18

### Fixes

- Only capture mouse events for changing the carousel when there are at least two attachments.


src/icons.js
import React from 'react'

export default {
    image: (props) => (
        <svg {...props} viewBox="0 0 100 100" version="1.1" xmlns="http://www.w3.org/2000/svg"
             style={{shapeRendering: 'geometricprecision'}}>
            <path
                d="M92.1065787,11 C96.4599604,11 100.001111,14.5411504 99.9999997,18.8945321 L99.9999997,82.0530115 C99.9999997,86.404171 96.4599604,89.9464325 92.1065787,89.9464325 L7.89453214,89.9464325 C3.54226157,89.9464325 0,86.4063932 0,82.0530115 L0,18.8945321 C0,14.5411504 3.54226157,11 7.89453214,11 L92.1065787,11 Z M76.9775218,54.7249302 L94.7366079,73.5595838 L94.7366079,18.8945321 C94.7366079,17.4422938 93.5588171,16.2622807 92.1065787,16.2622807 L7.89453214,16.2622807 C6.44340491,16.2622807 5.2633918,17.4422938 5.2633918,18.8945321 L5.2633918,69.2239801 L27.1814131,45.3126034 C28.3414259,44.0459226 29.9036655,43.3492482 31.5792397,43.3492482 C33.2548139,43.3492482 34.8170534,44.0459226 35.9770663,45.3126034 L56.2050688,67.3795152 L68.1352013,54.7249302 C69.3063254,53.4838053 70.8763429,52.8004643 72.5563615,52.8004643 C74.2363802,52.8004643 75.8052865,53.4838053 76.9775218,54.7249302 Z M75.2630583,45.21038 C70.0310226,45.21038 65.7896197,40.9689771 65.7896197,35.7369415 C65.7896197,30.5049058 70.0310226,26.2635029 75.2630583,26.2635029 C80.4950939,26.2635029 84.7364969,30.5049058 84.7364969,35.7369415 C84.7364969,40.9689771 80.4950939,45.21038 75.2630583,45.21038 Z"
                fill="currentColor"
                fillRule="nonzero"
            />
        </svg>
    )
}

src/CoverField.js
import React from 'react'
import ReactDOM from 'react-dom'
import {css} from 'emotion'
import times from 'lodash/times'
import icons from './icons'

const AttachmentContainer = ({children, innerWidth, offset}) => (
    <div
        className={css`
            width: 100%;
            overflow: hidden;
            position: relative;
            transition: left 200ms ease 0s;
        `}
        style={{
            height: 180
        }}
    >
        <div
            className={css`
                position: absolute;
                display: flex;
            `}
            style={{
                height: 180,
                width: innerWidth,
                left: 0 - offset
            }}
        >
            {children}
        </div>
    </div>
)

const Attachment = ({url, width, coverFitTypeId}) => (
    <div
        className={css`
            height: 180px;
            overflow: hidden;
            background-position: center center;
            background-size: ${coverFitTypeId === 'cover' ? 'cover' : 'contain'};
            background-repeat: no-repeat;
            background-image: url(${url});
        `}
        style={{
            width
        }}
    />
)

const Indicator = ({active}) => (
    <div
        className={css`
            flex: 1 1 auto;
            min-width: 0;
            min-height: 0;
            padding: 2px;
            display: flex;
        `}
    >
        <div
            className={css`
                flex: 1 1 auto;
                min-width: 0;
                min-height: 0;
                border-radius: 9999px;
                background-color: ${active ? 'hsl(0,0%,100%)' : 'hsla(0,0%,100%,0.5)'};
            `}
        />
    </div>
)

const Indicators = ({active, index, total}) => (
    <div
        className={css`
            position: absolute;
            bottom: 2px;
            left: 2px;
            right: 2px;
            height: 6px;
            opacity: ${active ? '1' : '0'};
            transform: ${active ? 'translateY(0px)' : 'translateY(8px)'};
            display: flex;
            transition: .085s all ease-in;
            background-color: hsla(0,0%,0%,0.25);
            border-radius: 9999px;
        `}
    >
        {times(total).map(i => (
            <Indicator
                key={i}
                active={index === i}
            />
        ))}
    </div>
)

export default class CoverField extends React.Component {

    index = 0

    state = {
        index: 0,
        hover: false,
        width: null
    }

    componentDidMount() {

        const el = ReactDOM.findDOMNode(this)

        const width = el.clientWidth

        this.setState({
            width
        })
    }

    render() {

        const {height, coverFitTypeId, attachments} = this.props

        const {width, index, hover} = this.state

        const attachEvents = attachments && attachments.length > 1

        return (
            <div
                className={css`
                    width: 100%;
                    position: relative;
                    background-color: #f9f9f9;
                `}
                style={{
                    height
                }}
                onMouseEnter={attachEvents ? this.handleMouseEnter : null}
                onMouseLeave={attachEvents ? this.handleMouseLeave : null}
                onMouseMove={attachEvents ? this.handleMouseMove : null}
            >
                {attachments && attachments.length ? (
                    <div
                        className={css`
                            width: 100%;
                            height: 100%;
                        `}
                    >
                        <AttachmentContainer
                            offset={index * width}
                            innerWidth={attachments.length * width}
                        >
                            {attachments.map((attachment, index) => (
                                <Attachment
                                    key={index}
                                    url={attachment.thumbnails.medium.url}
                                    width={width}
                                    coverFitTypeId={coverFitTypeId}
                                />
                            ))}
                        </AttachmentContainer>
                        {attachments.map((attachment, index) => (
                            <link key={index} rel={'prefetch'} href={attachment}/>
                        ))}
                        {attachments.length > 1 ? (
                            <Indicators
                                active={hover}
                                index={index}
                                total={attachments.length}
                            />
                        ) : null}
                    </div>
                ) : (
                    <div
                        className={css`
                            width: 100%;
                            height: 100%;
                            display: flex;
                            align-items: center;
                            justify-content: center;
                        `}
                    >
                        {icons.image({width: 100, color: '#f0f0f0'})}
                    </div>
                )}
            </div>
        )
    }

    handleMouseEnter = () => {

        if (!this.props.attachments) return
        if (this.props.attachments.length < 2) return

        this.setState({
            hover: true
        })
    }

    handleMouseMove = (e) => {

        const params = {
            targetWidth: e.currentTarget.clientWidth,
            targetOffset: e.currentTarget.offsetLeft,
            clientOffset: e.nativeEvent.clientX
        }

        const {targetWidth, targetOffset, clientOffset} = params

        const offset = clientOffset - targetOffset
        const diff = offset / targetWidth
        const total = this.props.attachments.length

        const index = Math.floor(total * diff)


        if (index !== this.index) {
            this.setIndex(index)
        }
    }

    setIndex = index => {
        this.index = index
        this.setState({
            index
        })
    }

    handleMouseLeave = () => {
        this.setIndex(0)
        this.setState({
            hover: false
        })
    }
}

src/RecordTitle.js
import React from 'react'
import {css} from 'emotion'

const RecordTitle = ({children}) => (
    <div
        className={css`
            display: flex;
            align-items: center;
            margin-top: 8px;
            padding-left: 16px;
            padding-right: 16px;
            font-size: 16px;
        `}
    >
        <div
            className={css`
                font-weight: 600;
                max-width: 100%;
                overflow: hidden;
                text-overflow: ellipsis;
                white-space: nowrap;
                color: #000;
                flex: 1 1 auto;
                min-width: 0;
                min-height: 0;
                line-height: 1.5;
            `}
        >
            {children}
        </div>
    </div>
)

export default RecordTitle

src/index.js
import React from 'react'
import PropTypes from 'prop-types'
import {css, cx} from 'emotion'
import CoverField from './CoverField'
import RecordTitle from './RecordTitle'
import calculateCardHeight from './utils/calculateCardHeight'

const PRIMARY_FIELD_HEIGHT = 32
const FIELD_WRAP_HEIGHT = 31 // spacing around a field
const CARD_BOTTOM_PADDING = 10

const THUMBNAIL_SHAPE = PropTypes.shape({
    height: PropTypes.number,
    width: PropTypes.number,
    url: PropTypes.string.isRequired
})

class RecordGalleryCard extends React.Component {

    static propTypes = {
        id: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        coverAttachments: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.string.isRequired,
                filename: PropTypes.string.isRequired,
                size: PropTypes.number,
                type: PropTypes.oneOf([
                    'image/jpeg',
                    'image/jpg',
                    'image/png',
                    'image/gif',
                    'audio/mpeg',
                    'video/mp4',
                    'video/ogg'
                ]).isRequired,
                thumbnails: PropTypes.shape({
                    small: THUMBNAIL_SHAPE,
                    medium: THUMBNAIL_SHAPE,
                    large: THUMBNAIL_SHAPE,
                })
            })
        ),
        coverFitTypeId: PropTypes.oneOf(['cover', 'fit']),
        coverEnabled: PropTypes.bool,
        coverHeight: PropTypes.number,
        fields: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.string.isRequired,
                name: PropTypes.string.isRequired
            })
        ),
        visibleFieldOrder: PropTypes.arrayOf(
            PropTypes.string.isRequired
        ),
        fieldHeightGetter: PropTypes.func,
        fieldRenderer: PropTypes.func
    }

    static defaultProps = {
        visibleFieldOrder: [],
        coverHeight: 180,
        coverEnabled: false,
        coverFitTypeId: 'cover'
    }

    render() {

        const {
            id,
            name,
            fieldRenderer,
            fieldHeightGetter,
            coverAttachments,
            coverEnabled,
            coverFitTypeId,
            coverHeight,
            visibleFieldOrder,
            onClick
        } = this.props

        const fieldsById = this.props.fields.reduce((result, field) => {
            result[field.id] = field
            return result
        }, {})

        const fields = visibleFieldOrder.map(id => {
            return fieldsById[id]
        })

        return (
            <div
                className={cx(
                    css`
                        background-color: #fff;
                        user-select: none;
                        width: 100%;
                        cursor: pointer;
                        border-radius: 6px;
                        overflow: hidden;
                        padding-bottom: 10px;
                        box-shadow: 0 0 0 1px rgba(114,121,133,.3);
                    `,
                    onClick ? css`
                    &:hover {
                        box-shadow: 0 0 0 1px rgba(114,121,133,.5);
                    }` : null
                )}
                onClick={e => {

                    if (onClick) {
                        onClick({
                            e,
                            id
                        })
                    }
                }}
            >
                {coverEnabled ? (
                    <CoverField
                        coverFitTypeId={coverFitTypeId}
                        attachments={coverAttachments}
                        height={coverHeight}
                    />
                ) : null}
                <RecordTitle>
                    {name}
                </RecordTitle>
                {fields.map((field, index) => {

                    const height = fieldHeightGetter({field})

                    return (
                        <div
                            key={field.id}
                            className={css`
                                padding-left: 16px;
                                padding-right: 16px;
                                padding-top: 8px;
                                padding-bottom: 4px;
                            `}
                        >
                            <div
                                className={css`
                                    margin-bottom: 4px;
                                    display: flex;
                                    align-items: center;
                                    color: #000;
                                `}
                            >
                                <div
                                    className={css`
                                        flex: 1 1 auto;
                                        min-width: 0;
                                        min-height: 0;
                                        font-weight: 600;
                                        max-width: 100%;
                                        overflow: hidden;
                                        text-overflow: ellipsis;
                                        white-space: nowrap;
                                        font-size: 12px;
                                    `}
                                >
                                    {field.name}
                                </div>
                            </div>
                            <div
                                style={{
                                    height
                                }}
                            >
                                {fieldRenderer({
                                    id,
                                    field,
                                    index,
                                    props: {
                                        id: field.id,
                                        contextId: 'recordGalleryCard',
                                        roleId: 'readOnly'
                                    }
                                })}
                            </div>
                        </div>
                    )
                })}
            </div>
        )
    }
}

RecordGalleryCard.calculateCardHeight = calculateCardHeight

export default RecordGalleryCard